jax_galsim/moffat.py

```diff
@@ -0,0 +1,467 @@
+
+@jax.jit
+def _Knu(nu, x):
+    """Modified Bessel 2nd kind for Untruncated Moffat"""
+    return tfp.substrates.jax.math.bessel_kve(nu * 1.0, x) / jnp.exp(jnp.abs(x))
+
+
+@jax.jit
+def MoffatIntegrant(x, k, beta):
+    """For truncated Hankel used in truncated Moffat"""
+    return x * jnp.power(1 + x**2, -beta) * j0(k * x)
+
+
+def _xMoffatIntegrant(k, beta, rmax, quad):
+    return quad_integral(partial(MoffatIntegrant, k=k, beta=beta), 0.0, rmax, quad)
+
+
+@jax.jit
+def _hankel(k, beta, rmax):
+    return jax.vmap(
+        partial(
+            _xMoffatIntegrant, beta=beta, rmax=rmax, quad=ClenshawCurtisQuad.init(150)
+        )
+    )(k)
+
+
+def MoffatCalculateSRFromHLR(re, rm, beta, Nloop=1000):
+    """
+    The basic equation that is relevant here is the flux of a Moffat profile
+    out to some radius.
+
+    flux(R) = int( (1+r^2/rd^2 )^(-beta) 2pi r dr, r=0..R )
+            = (pi rd^2 / (beta-1)) (1 - (1+R^2/rd^2)^(1-beta) )
+    For now, we can ignore the first factor.  We call the second factor fluxfactor below,
+    or in this function f(R).
+    We are given two values of R for which we know that the ratio of their fluxes is 1/2:
+    f(re) = 0.5 * f(rm)
+
+    nb1. rd aka r0 aka the scale radius
+    nb2. In GalSim definition rm = 0 (ex. no truncated Moffat) means in reality rm=+Inf.
+         BUT the case rm==0 is already done, so HERE rm != 0
+    """
+    # JEC Where to do these checking?
+    # assert rm != 0.0, f"MoffatCalculateSRFromHLR: rm=={rm} should be done elsewhere"
+    #
+    # assert (
+    #    rm > jnp.sqrt(2.0) * re
+    # ), f"MoffatCalculateSRFromHLR: Cannot find a scaled radius: rm={rm}, sqrt(2)*re={jnp.sqrt(2.) * re}"
+
+    ## fix loop iteration is faster and reach eps=1e-6 (single precision)
+    def body(i, xcur):
+        x = (1 + jnp.power(1 + (rm / xcur) ** 2, 1 - beta)) / 2
+        x = jnp.power(x, 1 / (1 - beta))
+        x = jnp.sqrt(x - 1)
+        return re / x
+
+    rd = jax.lax.fori_loop(0, Nloop, body, re)
+
+    return rd
+
+
+@_wraps(_galsim.Moffat)
+@register_pytree_node_class
+class Moffat(GSObject):
+    _is_axisymmetric = True
+    _is_analytic_x = True
+    _is_analytic_k = True
+
+    def __init__(
+        self,
+        beta,
+        scale_radius=None,
+        half_light_radius=None,
+        fwhm=None,
+        trunc=0.0,
+        flux=1.0,
+        gsparams=None,
+    ):
+        # notice that trunc==0. means no truncated Moffat.
+
+        gsparams = GSParams.check(gsparams)
+
+        # let define beta_thr a threshold to trigger the truncature
+        self._beta_thr = 1.1
+
+        # JEC Where to do these checking?
+        # if trunc == 0.0 and beta <= _beta_thr:
+        #    raise _galsim.GalSimRangeError(
+        #        f"Moffat profiles with beta <= {_beta_thr} must be truncated",
+        #        beta,
+        #        _beta_thr,
+        #    )
+        # if trunc < 0.0:
+        #    raise _galsim.GalSimRangeError("Moffat trunc must be >= 0", trunc, 0.0)
```

Comment:
same with this, see my other comment above about what to do with these checks

Reply:
done